qualtran/bloqs/arithmetic/bitwise.py

```diff
@@ -118,3 +123,73 @@ def _cxork() -> XorK:
     cxork = XorK(QUInt(8), 0b01010111).controlled()
     assert isinstance(cxork, XorK)
     return cxork
+
+
+@frozen
+class Xor(SpecializedSingleQubitControlledGate):
+    """Xor the value of one register into another via CNOTs.
+
+    When both registers are in computational basis and the destination is 0,
+    effectively copies the value of the source into the destination.
+
+    Args:
+        dtype: Data type of the input registers `x` and `y`.
+        control_val: An optional single bit control, apply the operation when
+            the control qubit equals the `control_val`.
+
+    Registers:
+        x: The source register.
+        y: The target register.
+    """
+
+    dtype: QDType
+    control_val: Optional[int] = None
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(x=self.dtype, y=self.dtype)
+
+    @cached_property
+    def control_registers(self) -> Tuple[Register, ...]:
+        if self.control_val is not None:
+            return (Register('ctrl', QBit()),)
+        return ()
+
+    def build_composite_bloq(self, bb: BloqBuilder, x: Soquet, y: Soquet) -> Dict[str, SoquetT]:
+        if not isinstance(self.dtype.num_qubits, int):
+            raise DecomposeTypeError("`dtype.num_qubits` must be a concrete value.")
+
+        xs = bb.split(x)
+        ys = bb.split(y)
+
+        for i in range(len(xs)):
+            xs[i], ys[i] = bb.add_t(CNOT(), ctrl=xs[i], target=ys[i])
+
+        return {'x': bb.join(xs), 'y': bb.join(ys)}
```

Comment:
I think you need to specify the `dtype` to `bb.join()` or else it'll join to `QAny` and the output soquet dtype would be different from expected dtype. I'm surprised this doesn't raise an error already.